File: vortexdbg-android/src/main/java/com/vortexdbg/linux/signal/SigAction.java
package com.vortexdbg.linux.signal;

import com.vortexdbg.Emulator;
import com.vortexdbg.pointer.VortexdbgStructure;
import com.sun.jna.Pointer;

public abstract class SigAction extends VortexdbgStructure {

    private static final int SA_SIGINFO = 0x00000004;

    public static SigAction create(Emulator<?> emulator, Pointer ptr) {
        if (ptr == null) {
            return null;
        }
        SigAction action;
        if (emulator.is32Bit()) {
            action = new SigAction32(ptr);
        } else {
            action = new SigAction64(ptr);
        }
        action.unpack();
        return action;
    }

    public abstract long getSaHandler();
    public abstract void setSaHandler(long sa_handler);

    public abstract long getSaRestorer();
    public abstract void setSaRestorer(long sa_restorer);

    public boolean needSigInfo() {
        return (getFlags() & SA_SIGINFO) != 0;
    }

    public abstract long getMask();

    public abstract void setMask(long mask);

    public abstract int getFlags();

    public abstract void setFlags(int flags);

    public SigAction(Pointer p) {
        super(p);
    }

}

File: vortexdbg-android/src/main/kotlin/com/vortexdbg/linux/signal/SigAction.kt
package com.vortexdbg.linux.signal

import com.vortexdbg.Emulator
import com.vortexdbg.pointer.VortexdbgStructure
import com.sun.jna.Pointer

abstract class SigAction(p: Pointer?) : VortexdbgStructure(p) {

    abstract fun getSaHandler(): Long
    abstract fun setSaHandler(sa_handler: Long)

    abstract fun getSaRestorer(): Long
    abstract fun setSaRestorer(sa_restorer: Long)

    open fun needSigInfo(): Boolean {
        return (getFlags() and SA_SIGINFO) != 0
    }

    abstract fun getMask(): Long

    abstract fun setMask(mask: Long)

    abstract fun getFlags(): Int

    abstract fun setFlags(flags: Int)

    companion object {
        private const val SA_SIGINFO = 0x00000004

        @JvmStatic
        fun create(emulator: Emulator<*>, ptr: Pointer?): SigAction? {
            if (ptr == null) {
                return null
            }
            val action: SigAction = if (emulator.is32Bit()) {
                SigAction32(ptr)
            } else {
                SigAction64(ptr)
            }
            action.unpack()
            return action
        }
    }

}
